### PR TITLE
LVGL using standard HAL TFT IO (SPI and FSMC)

### PR DESCRIPTION
**Pull request by rhapsodyv**
### Description

We have a lot of TFT/XPT IO repeated code in STM32F1. This PR begins to change that. 

Now LVGL uses the standard HAL TFT IO - FSMC and SPI, and XPT2046. 

Standard HAL TFT and TOUCH IO are in:
```
#include HAL_PATH(../../HAL, tft/xpt2046.h) 
#include HAL_PATH(../../HAL, tft/tft_spi.h)
#include HAL_PATH(../../HAL, tft/tft_fsmc.h)
```

Next PR I will see to migrate emulated DOGM.

### Benefits

- less repeated code
- more performance (HAL TFT SPI uses 16 bit DMA, the old LVGL SPI DMA were 8 bit, so it became very fast now)
- less errors
- more easily to port LVGL to others HAL

### Related Issues

#18958 
#18952

### CODE REVIEW

**Comment by looxonline** [comment]
Yes. Waiting for this to be used in a LP1678 SPI port.

**Comment by rhapsodyv** [comment]
> Yes. Waiting for this to be used in a LP1678 SPI port.

Which board?

**Comment by looxonline** [comment]
> > Yes. Waiting for this to be used in a LP1678 SPI port.
> 
> Which board?

Have not checked DMA access on the SPI port exposed on the EXP header of the SKR1.4 but if there is access then it would be great to use it with that board.

**Comment by rhapsodyv** [comment]
> > > Yes. Waiting for this to be used in a LP1678 SPI port.
> > 
> > 
> > Which board?
> 
> Have not checked DMA access on the SPI port exposed on the EXP header of the SKR1.4 but if there is access then it would be great to use it with that board.

I wil test it. But to run LVGL UI, you need access to spi flash too. The SKR 1.4 board only have spi flash in the TFT, not in the board. I dont know if it expose too.

**Comment by looxonline** [comment]
> > > > Yes. Waiting for this to be used in a LP1678 SPI port.
> > > 
> > > 
> > > Which board?
> > 
> > 
> > Have not checked DMA access on the SPI port exposed on the EXP header of the SKR1.4 but if there is access then it would be great to use it with that board.
> 
> I wil test it. But to run LVGL UI, you need access to spi flash too. The SKR 1.4 board only have spi flash in the TFT, not in the board. I dont know if it expose too.

Yes, there is an exposed SPI header which is shared with the SD card. P0.26, P0.7, P0.8, and P0.9. I believe that these could easily be interfaced with a cheap, external W25Q32.

**Comment by rhapsodyv** [comment]
@thinkyhead  I have finished the emulated DOGM to use the same HAL TFT IO classes.

Now, we have only one FSMC and SPI TFT in STM32F1, that follow this interface:
```cpp
#include HAL_PATH(../../HAL, tft/tft_spi.h)
#include HAL_PATH(../../HAL, tft/tft_fsmc.h)
```
And it is already available in STM32.

Any other HAL that wants TFT, just need implement this two classes, and LVGL + emulated dogm will work!


Do I commit it here? Or wait the merge and do another PR, specific for dogm?

**Comment by rhapsodyv** [comment]
LVGL, Emulated DOGM and the new Color UI, now all are using the same low level IO abstraction. To use any of those UI, the HAL just need the low level TFT IO classes. STM32 and STM32F1 already have it, so all 3 UI are available for them.

**Comment by rhapsodyv** [comment]
@thinkyhead 
This port emulated DOGM to use the same tft io: https://github.com/rhapsodyv/Marlin/commit/6bd3d04b09c8b5c1323b5ba0153536351d2366a6

Do I push it here? Or do you prefere separated PR to better organize merges?

After that commit, we have only two things to finish that refactoring:
1) I will join all xpt code
2) we can make a cleanup of the defines LCD_ and TFT_ defines

Lots of repeated code removed!!

File: Marlin/src/lcd/extui/lib/mks_ui/tft_lvgl_configuration.cpp [hunk 1]
```diff
@@ -556,174 +513,23 @@ unsigned int getTickDiff(unsigned int curTick, unsigned int lastTick) {
...
+    *y = int16_t((int32_t(*y) * XPT2046_Y_CALIBRATION) >> 16) + XPT2046_Y_OFFSET;
   }
 
-  *Y_Addata = (y_addata[times / 2 - 1] + y_addata[times / 2]) / 2;
-}
-
-#define ADC_VALID_OFFSET  10
+  #if ENABLED(GRAPHICAL_TFT_ROTATE_180)
+    x = 480 - x;
+    y = 320 - y;
```
Comment (thinkyhead) [inline Marlin/src/lcd/extui/lib/mks_ui/tft_lvgl_configuration.cpp]: The 480 and 320 here should probably be `LCD_FULL_PIXEL_WIDTH`/`HEIGHT` and in general any mysterious numbers used more than a few times should be given named defines.

Comment (rhapsodyv) [reply Marlin/src/lcd/extui/lib/mks_ui/tft_lvgl_configuration.cpp]: Fixed

**Comment by thinkyhead** [comment]
Have a look at `lcd/scaled_tft.h` and see where it makes sense to simply set `FSMC_UPSCALE` to 3 in the pins files. The `lcd/scaled_tft.h` header is included by some files, so check where it is now included and consider where it makes sense in relation to this PR to `#include` it as an alternative to defining `LCD_FULL_PIXEL_WIDTH`/`HEIGHT`.

If there are new or other conditions that pertain to the values of `LCD_FULL_PIXEL_WIDTH`/`HEIGHT`, then we can either extend that header, add a new one, or look at whether it should be defined in `HAL/STM32/inc/Conditionals_post.h` instead.

**Comment by rhapsodyv** [comment]
> Have a look at `lcd/scaled_tft.h` and see where it makes sense to simply set `FSMC_UPSCALE` to 3 in the pins files. The `lcd/scaled_tft.h` header is included by some files, so check where it is now included and consider where it makes sense in relation to this PR to `#include` it as an alternative to defining `LCD_FULL_PIXEL_WIDTH`/`HEIGHT`.
> 
> If there are new or other conditions that pertain to the values of `LCD_FULL_PIXEL_WIDTH`/`HEIGHT`, then we can either extend that header, add a new one, or look at whether it should be defined in `HAL/STM32/inc/Conditionals_post.h` instead.

Right now, the LVGL just works in 480x320. It's the reason of that lot of hardcoded numbers 480/320. MKS is working in a 320x240 version of the LVGL UI, but they didn't send anything yet. And it is the reason that it defines it to 480/320 if it is not defined, too.

Specific for this PR, the `lcd / scaled_tft.h` and` FSMC_UPSCALE` are not used.

But, for the next one, that migrates emulated DOGM to use TFT_IO... yes, we will have some fun work  _removing and standardising_ some LCD / TFT defines. I think all LCD_ PIXEL* defines may die, in favor of TFT_ * ones, as all TFT IO will be handled by `HAL/tft/tft_(fsmc/spi).h`.

**Comment by rhapsodyv** [comment]
I think is better have two PR, this one for LVGL (that is more specific for some board and tft), and a new for emulated DOGM, that are used for a lot more boards and tfts. 
And I think you will want to refactor a lot of defines, removing duplicated ones and better organise it. 

The current state is: we have SPI_TFT* (mks use), we have LCD_PIXEL (dogm and a few use), we have TFT_* that new color UI use... 
After this work we are doing, we can have only one set of defines, as ALL ui will use the same classes for TFT IO!

**Comment by thinkyhead** [comment]
Ok, it will be good to wrangle these at an early stage.

I'm following the patterns to see how these all relate.

image

It's containable. I see from this that the `FSMC_UPSCALE` pertains to how the screen will be drawn and how touches will be interpreted, but also that `FSMC_UPSCALE` is not tied specifically to the pixel-size. One should be able to decide to do 2x or even 1x scaling on a 480x320 screen and make it work, after all….

**Comment by rhapsodyv** [comment]
> It's containable. I see from this that the `FSMC_UPSCALE` pertains to how the screen will be drawn and how touches will be interpreted, but also that `FSMC_UPSCALE` is not tied specifically to the pixel-size. One should be able to decide to do 2x or even 1x scaling on a 480x320 screen and make it work, after all….

Sure, after this refactoring, we can handle upscale at a low level, making it available for all UI.

**Pull request merged**